src/core/mac/mac_frame.cpp

```diff
@@ -153,9 +153,17 @@ void TxFrame::Info::PrepareHeadersIn(TxFrame &aTxFrame) const
                 break;
             }
         }
-        else if (mPanIds.GetSource() != mPanIds.GetDestination()) // Rows 9-14.
+        else if (mPanIds.IsDestinationPresent())
```

Comment:
I am not sure this is correct. Can you please point to specification?

The table (which I believe is copied from the text) is unfortunately not clear and can be misleading. It represents what is included in the frame and has the following associated footnote:

Check here: https://mentor.ieee.org/802.15/dcn/15/15-15-0911-01-0mag-proper-pan-id-field-settings-for-802-15-4-2015.docx
> If both the destination and source addressing information is present and either is a short address, the MAC sublayer shall compare the destination and source PAN IDs and the PAN ID Compression field shall be set to zero if and only if the PAN identifiers are identical.

When we have both source and destination addresses and at least one is "short," then I believe both source/destination PAN IDs are also expected. Then the "Pan ID compression" flag is determined based on whether the two PAN IDs match, and then one is omitted from the encoded frame. I think it does not mean we did not have this PAN ID when we were preparing the frame; it should be interpreted as it was optimized (compressed) as it was the same as the other destination PAN ID.

Reply:
@abtink  When I added the following test cases to the `TestMacHeader()` of `tests/unit/test_mac_frame.cpp`, I found these test cases are failed. These test cases meet rules of row 12 to 14 of generating the version 2015 frame. 

```c++
         {kVer2015, kExtdAddr, kNoPanId, kShrtAddr, kUsePanId1, kNoSec, kModeId1, 17, 2}, // row 12
         {kVer2015, kShrtAddr, kNoPanId, kExtdAddr, kUsePanId1, kNoSec, kModeId1, 17, 2}, // row 13
         {kVer2015, kShrtAddr, kNoPanId, kShrtAddr, kUsePanId1, kNoSec, kModeId1, 11, 2}, // row 14
```

For these test cases, no source PanIds are present. But when paring the frame, the source PanId is present. I looked at how the 2015 frame is generated. I found if the destination PanId is present and the source PanId is not present, a zero-valued source PanId will be allocated for the frame. I think this is not correct.

```c++

tests/unit/test_mac_frame.cpp +393:

          if (frame.IsSrcPanIdPresent())
          {
              SuccessOrQuit(frame.GetSrcPanId(panId));
              VerifyOrQuit(panId == frameInfo.mPanIds.GetSource());
              VerifyOrQuit(frameInfo.mPanIds.IsSourcePresent());   <----failed
          }
```
If we follow your assumption `I believe both source/destination PAN IDs are also expected`, the rule of row 12 to 14 do not exist, right? I am not sure if test cases for row 12 to 14 are intentionally missing.

Reply:
@zhanglongxia, I think the new test cases you mentioned may not be valid or allowed:

```
{kVer2015, kExtdAddr, kNoPanId, kShrtAddr, kUsePanId1, kNoSec, kModeId1, 17, 2}, // row 12
{kVer2015, kShrtAddr, kNoPanId, kExtdAddr, kUsePanId1, kNoSec, kModeId1, 17, 2}, // row 13
{kVer2015, kShrtAddr, kNoPanId, kShrtAddr, kUsePanId1, kNoSec, kModeId1, 11, 2}, // row 14
```

My understanding is that if we use a short address for either the source or destination, it needs to be associated with a PAN ID. A short address seems to be meaningful only within a PAN, so it needs to be associated with a PAN ID.

The way to interpret the table is to look at what is encoded in the generated MAC header and not the input used to generate the frame. The table's main purpose is to determine the "PAN ID Compression" flag and show which combinations of flags are valid and can appear in the MAC header.

The entries in rows 9-14 in the table have a `1` footnote in the text that says:

> `1` If both the destination and source addressing information is present and either is a short address, the MAC sublayer shall compare the destination and source PAN IDs and the PAN ID Compression field shall be set to zero if and only if the PAN identifiers are identical.


So rules 9-14 are applied for these sets of cases. In one set, we use `kUsePanId1` for the source and `kUsePanId2` for the destination. In the other set, we use `kUsePanId1` for both and expect "PAN ID compression" to be applied and elide the encoded source PAN ID field in the header:

```
{kVer2015, kShrtAddr, kUsePanId1, kShrtAddr, kUsePanId2, kNoSec, kModeId1, 11, 2},
{kVer2015, kShrtAddr, kUsePanId1, kExtdAddr, kUsePanId2, kNoSec, kModeId1, 17, 2},
{kVer2015, kExtdAddr, kUsePanId1, kShrtAddr, kUsePanId2, kNoSec, kModeId1, 17, 2},

{kVer2015, kShrtAddr, kUsePanId1, kShrtAddr, kUsePanId1, kNoSec, kModeId1, 9, 2},
{kVer2015, kShrtAddr, kUsePanId1, kExtdAddr, kUsePanId1, kNoSec, kModeId1, 15, 2},
{kVer2015, kExtdAddr, kUsePanId1, kShrtAddr, kUsePanId1, kNoSec, kModeId1, 15, 2},
```

Reply:
@abtink Thanks for your detailed explanation. Could you please help add more comments where the 2015 frame is generated to avoid confusing readers like me?

Reply:
We can add the same footnote from the spec below the table. 
Beyond this, do you have any suggestions?

Reply:
Check https://github.com/openthread/openthread/pull/10914.